### PR TITLE
Cleanup loops and ranges

### PR DESCRIPTION
**Pull request by oscargus**
#### Brief description of what is fixed or changed

Removed reundant zeros in ranges

Replaced range-loops with enumerate, zip, and/or loop comprehensions.

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/combinatorics/testutil.py [hunk 1]
```diff
@@ -346,12 +343,12 @@ def graph_certificate(gr):
...
-        n = vlen[i]
-        if n:
-            base, gens = get_symmetric_group_sgs(i)
-            v.append((base, gens, n, 0))
+    v = [(*get_symmetric_group_sgs(i), n, 0) for i, n in enumerate(vlen) if n]
+    #for i in range(len(vlen)):
+    #    n = vlen[i]
+    ##    if n:
+    #        base, gens = get_symmetric_group_sgs(i)
+    #        v.append((base, gens, n, 0))
```
Comment (oscargus) [inline sympy/combinatorics/testutil.py:351]: This should be reverted...

File: sympy/physics/quantum/gate.py [hunk 2]
```diff
@@ -245,9 +245,9 @@ def _apply_operator_Qubit(self, qubits, **options):
             # Make a copy of the incoming qubits.
             new_qubit = qubits.__class__(*qubits.args)
             # Flip the bits that need to be flipped.
-            for bit in range(len(targets)):
-                if new_qubit[targets[bit]] != (index >> bit) & 1:
-                    new_qubit = new_qubit.flip(targets[bit])
+            for bit, target in enumerate(targets):
+                if new_qubit[target != (index >> bit) & 1:
```
Comment (oscargus) [inline sympy/physics/quantum/gate.py:249]: Typo

File: sympy/logic/algorithms/dpll2.py [hunk 3]
```diff
@@ -143,21 +143,19 @@ def _initialize_clauses(self, clauses):
...
-            if 1 == len(self.clauses[i]):
-                self._unit_prop_queue.append(self.clauses[i][0])
+            if 1 == len(clause):
+                self._unit_prop_queue.append(clause[0])
                 continue
 
-            self.sentinels[self.clauses[i][0]].add(i)
-            self.sentinels[self.clauses[i][-1]].add(i)
+            self.sentinels[self.clause[0]].add(i)
+            self.sentinels[self.clause[-1]].add(i)
```
Comment (oscarbenjamin) [inline sympy/logic/algorithms/dpll2.py:156]:
```suggestion
            self.sentinels[clause[0]].add(i)
            self.sentinels[clause[-1]].add(i)
```

File: sympy/holonomic/holonomic.py [hunk 4]
```diff
@@ -330,7 +330,7 @@ def __pow__(self, n):
         # if self is `Dx`
         if self.listofpoly == self.parent.derivative_operator.listofpoly:
             sol = []
-            for i in range(0, n):
+            for _ in range(n):
```
Comment (smichr) [inline sympy/holonomic/holonomic.py:333]: would it make sense to just do `sol = [self.parent.base.zero]*n`?

File: sympy/holonomic/holonomic.py [hunk 5]
```diff
@@ -2186,11 +2186,11 @@ def from_hyper(func, x0=0, evalf=False):
 
     # generalized hypergeometric differential equation
     r1 = 1
-    for i in range(len(a)):
-        r1 = r1 * (x * Dx + a[i])
+    for ai in a:
+        r1 = r1 * (x * Dx + ai)
```
Comment (smichr) [inline sympy/holonomic/holonomic.py:2190]: `xdx = x*Dx; r1 = Mul(*[xdx + ai for ai in a])` and similar for `r2` but also multiply by Dx when done.

Comment (smichr) [reply sympy/holonomic/holonomic.py:2190]: same for the other ones below. As I recall, building up the Add/Mul is a slow way to go; better to build args then do op all at once.

File: sympy/holonomic/holonomic.py [hunk 6]
```diff
@@ -330,7 +330,7 @@ def __pow__(self, n):
         # if self is `Dx`
         if self.listofpoly == self.parent.derivative_operator.listofpoly:
             sol = []
```
Comment (smichr) [inline sympy/holonomic/holonomic.py:332]:
```suggestion
            sol = [self.parent.base.zero]*n
```

File: sympy/holonomic/holonomic.py (hunk 4 above)
Comment (smichr) [inline sympy/holonomic/holonomic.py:333]:
```suggestion
```

File: sympy/holonomic/holonomic.py [hunk 7]
```diff
@@ -330,7 +330,7 @@ def __pow__(self, n):
...
-            for i in range(0, n):
+            for _ in range(n):
                 sol.append(self.parent.base.zero)
```
Comment (smichr) [inline sympy/holonomic/holonomic.py:334]:
```suggestion
```

File: sympy/holonomic/holonomic.py [hunk 8]
```diff
@@ -2263,17 +2259,10 @@ def from_meijerg(func, x0=0, evalf=False, initcond=True, domain=QQ):
...
-
-    r2 = 1
-
-    for i in range(len(b)):
-        r2 *= x * Dx - b[i]
-
+    xDx = x*Dx
+    xDx1 = xDx + 1
+    r1 = Mul(*([x*(-1)**(m + n - p)] + [xDx1 - ai for ai in a]))
+    r2 = Mul(*[xDx - bi for bi in b]
```
Comment (smichr) [inline sympy/holonomic/holonomic.py:2265]:
```suggestion
    r2 = Mul(*[xDx - bi for bi in b])
```

File: sympy/holonomic/holonomic.py [hunk 9]
```diff
@@ -2185,12 +2183,10 @@ def from_hyper(func, x0=0, evalf=False):
...
-    r1 = 1
-    for i in range(len(a)):
-        r1 = r1 * (x * Dx + a[i])
-    r2 = Dx
-    for i in range(len(b)):
-        r2 = r2 * (x * Dx + b[i] - 1)
+    xDx = x*Dx
+    xDx_1 = x*Dx - 1
+    r1 = Mul(*[xDx + ai for ai in a])
+    r2 = Mul(*[xDx_1 + bi for bi in b])
```
Comment (smichr) [inline sympy/holonomic/holonomic.py:2189]:
```suggestion
    r1 = Mul(*[xDx + ai for ai in a])
    xDx_1 = xDx - 1
    # r2 = Mul(*([Dx] + [xDx_1 + bi for bi in b]))  # XXX gives sympify error
    r2 = Dx
    for bi in b:
        r2 *= xDx_1 + bi
```

**Comment by smichr** [comment]
I don't understand why there is a sympify error in the holonomic code when trying to process all factors at once with Mul instead of building them up incrementally. I un-flattened one of them because of the sympify error from something like this:
```python
>>> from sympy.holonomic import *
>>> from_hyper(hyper([], [S(3)/2], x**2/4))
```

File: sympy/holonomic/holonomic.py [hunk 10]
```diff
@@ -2185,12 +2183,13 @@ def from_hyper(func, x0=0, evalf=False):
     R, Dx = DifferentialOperators(QQ.old_poly_ring(x), 'Dx')
 
     # generalized hypergeometric differential equation
-    r1 = 1
-    for i in range(len(a)):
-        r1 = r1 * (x * Dx + a[i])
+    xDx = x*Dx
+    r1 = Mul(*[xDx + ai for ai in a])
```
Comment (smichr) [inline sympy/holonomic/holonomic.py:2187]:
```suggestion
    r1 = 1
    for ai in a:  # XXX gives sympify error if Mul is used with list of all factors
        r1 *= xDx + ai
```

File: sympy/holonomic/holonomic.py [hunk 11]
```diff
@@ -2263,17 +2262,13 @@ def from_meijerg(func, x0=0, evalf=False, initcond=True, domain=QQ):
...
     # Meijer G-function.
-    mnp = (-1)**(m + n - p)
-    r1 = x * mnp
-
-    for i in range(len(a)):
-        r1 *= x * Dx + 1 - a[i]
-
+    xDx = x*Dx
+    xDx1 = xDx + 1
+    r1 = Mul(*([x*(-1)**(m + n - p)] + [xDx1 - ai for ai in a]))
```
Comment (smichr) [inline sympy/holonomic/holonomic.py:2267]:
```suggestion
    r1 = x*(-1)**(m + n - p)
    for ai in a:  # XXX gives sympify error if args given in list
        r1 *= xDx1 - ai
```

**Pull request merged**